runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -375,6 +428,11 @@ public RemoteBundle getBundle(
         }
       }
 
+      // Ensure client is referenced for this bundle, unref in close()
+      client.ref();
+      // Cleanup list of clients which were active during eviction but now do not hold references
+      evictedActiveClients.removeIf(c -> c.bundleRefCount.get() <= 0);
```

Comment:
I'm not sure I like this in the path for every bundle. We could probably move it below line 416 (`preparedClients.keySet().removeIf(c -> c.bundleRefCount.get() <= 0);`)

Reply:
This is a cheap operation. I don't think we can move it because we need to run this cleanup for all branches.

Reply:
I've improved this to only remove when environment expiration is used.